src/chmatch.c

```diff
@@ -24,26 +24,19 @@ static SEXP chmatchMain(SEXP x, SEXP table, int nomatch, bool chin, bool chmatch
   int *ansd = INTEGER(ans);
   if (!length(table)) { const int val=(chin?0:nomatch), n=LENGTH(x); for (int i=0; i<n; ++i) ansd[i]=val; UNPROTECT(1); return ans; }
   savetl_init();
-  const SEXP *xd = STRING_PTR(x);
   const int xlen = length(x);
+  int nprotect = 1;
+  SEXP ux = x;
   for (int i=0; i<xlen; i++) {
-    SEXP s = xd[i];
-    if (s != NA_STRING && ENC_KNOWN(s) != 64) { // PREV: s != NA_STRING && !ENC_KNOWN(s) - changed to fix for bug #5159. The previous fix
-                       // dealt with UNKNOWN encodings. But we could have the same string, where both are in different
-                       // encodings than ASCII (ex: UTF8 and Latin1). To fix this, we'll to resort to 'match' if not ASCII.
-                       // This takes care of all anomalies. It's unfortunate the 'chmatch' can't be used in these cases...
-                       // // fix for the 2nd part of bug #5159
-
-      // explanation for PREV case: (still useful to keep it here)
-      // symbols (i.e. as.name() or as.symbol()) containing non-ascii characters (>127) seem to be 'unknown' encoding in R.
-      // The same string in different encodings (where unknown encoding is different to known, too) are different
-      // CHARSXP pointers; this chmatch relies on pointer equality only. We tried mkChar(CHAR(s)) [ what install() does ]
-      // but this impacted the fastest cases too much. Hence fall back to match() on the first unknown encoding detected
-      // since match() considers the same string in different encodings as equal (but slower). See #2538 and #4818.
-      savetl_end();
-      UNPROTECT(1);
-      return (chin ? match_logical(table, x) : match(table, x, nomatch));
+    if (NEED2UTF8(STRING_ELT(x, i))) {
+      ux = PROTECT(allocVector(STRSXP, xlen)); ++nprotect;
+      for (int i=0; i<xlen; i++) SET_STRING_ELT(ux, i, ENC2UTF8(STRING_ELT(x, i)));
```

Comment:
not ideal to rely on scoping here, can change to a different loop index? or refactor not to nest the loops

Reply:
Thanks. Refactored.